### PR TITLE
fix bug where string slice flag is not assigned

### LINKED ISSUE #102621: string slice flag is not assigned
#### What happened:
`flags.Var(cliflag.NewStringSlice(&cloudConfig.Zones), "gce-zones", "")` ,  `&cloudConfig.Zones`  is not assigned

#### What you expected to happen:
`&cloudConfig.Zones`  is  assigned

#### How to reproduce it (as minimally and precisely as possible):
Use `cliflag.NewStringSlice` to parse the command line parameters

#### Anything else we need to know?:
I am preparing to merge pr to solve this problem

#### Environment:
- Kubernetes version (use `kubectl version`): N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by vazmin**
#### What type of PR is this?
/kind bug
/kind regression

#### What this PR does / why we need it:
string slice flag is not assigned

#### Which issue(s) this PR fixes:

Fixes #102621

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag_test.go [hunk 1]
```diff
@@ -83,8 +83,8 @@ func TestStringSlice(t *testing.T) {
 		} else if err != nil {
 			t.Errorf("%d: expected nil error, got %v", i, err)
 		}
-		if !reflect.DeepEqual(*v.value, test.expected) {
-			t.Errorf("%d: expected %+v, got %+v", i, test.expected, *v.value)
+		if !reflect.DeepEqual(s, test.expected) {
```
Comment (vazmin) [inline staging/src/k8s.io/component-base/cli/flag/string_slice_flag_test.go:86]: `s` is the flag, the expected should be compared with `s`

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go [hunk 2]
```diff
@@ -46,8 +46,7 @@ func (s *StringSlice) String() string {
 
 func (s *StringSlice) Set(val string) error {
 	if s.value == nil || !s.changed {
-		v := make([]string, 0)
-		s.value = &v
```
Comment (vazmin) [inline staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: `
func NewStringSlice(s *[]string) *StringSlice {
	return &StringSlice{value: s}
}
`
`s.value = &v`, this makes it use a new pointer, which obviously shouldn't be done

**Comment by leilajal** [comment]
/assign @jiahuif 
Would you please review this change. This is the fix to assign the zone values in cloud config. 
Thanks a lot.

**Review by leilajal — APPROVED** [review]

**Comment by leilajal** [comment]
/assign @sttts 
Thanks

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go (hunk 2 above)
Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: does 'flags.VarP()'  work?

Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: > does 'flags.VarP()' work?

https://github.com/kubernetes/kubernetes/blob/c9414c2bbaf9acfc245fc481ba2831efe7fdc7d6/staging/src/k8s.io/component-base/cli/flag/string_slice_flag_test.go#L69

 After change to `fs.VarP(v, "slice", "", "usage")`,  if the expected compared with `s`,  the test still fails

Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: `*s.value = make([]string, 0)` will panic if  s.Value is nil

Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: Do you mean this issues https://github.com/kubernetes/kubernetes/issues/100342?

`nil` judgment is added to this pr https://github.com/kubernetes/kubernetes/pull/100362 , so there will be no such problem.
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go#L41

Retest without problems.
```
make WHAT=test/e2e/e2e.test
./_output/bin/e2e.test --help
```

Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: how about change like that

```
func (s *StringSlice) Set(val string) error {
	if s.value == nil {
		return fmt.Errorf("return err")
	}
	if *s.value == nil || !s.changed {
		*s.value = make([]string, 0)
	}
	*s.value = append(*s.value, val)
	s.changed = true
	return nil
}
```

Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: > Do you mean this issues #100342?
> 
> `nil` judgment is added to this pr #100362 , so there will be no such problem.
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go#L41
> 
> Retest without problems.
> 
> ```
> make WHAT=test/e2e/e2e.test
> ./_output/bin/e2e.test --help
> ```

yes

Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: > how about change like that
> 
> ```
> func (s *StringSlice) Set(val string) error {
> 	if s.value == nil {
> 		return fmt.Errorf("return err")
> 	}
> 	if *s.value == nil || !s.changed {
> 		*s.value = make([]string, 0)
> 	}
> 	*s.value = append(*s.value, val)
> 	s.changed = true
> 	return nil
> }
> ```

There is already a nil judgment before, is this necessary? In fact, the original repaired pr only needs to be judged by nil, what I did is to restore the redundant changes. As you can see, one of the kind of pr is `regression`

Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: no, just  consider the Robustness of `Set` function, 
```
if err := NewStringSlice(nil).Set("value"); err != nil {
		t.Errorf("set value error: %v", err)
	}
````
the error is panic

Comment (stewart-yu) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: no, just  consider the Robustness of `Set` function, 
```
if err := NewStringSlice(nil).Set("value"); err != nil {
		t.Errorf("set value error: %v", err)
	}
````
the error is panic

Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: It looks good, but I think the role of StringSlice is to assign values to the input parameters, and it seems meaningless to pass nil.
:smile:

Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: > how about change like that
> 
> ```
> func (s *StringSlice) Set(val string) error {
> 	if s.value == nil {
> 		return fmt.Errorf("return err")
> 	}
> 	if *s.value == nil || !s.changed {
> 		*s.value = make([]string, 0)
> 	}
> 	*s.value = append(*s.value, val)
> 	s.changed = true
> 	return nil
> }
> ```

Modified and submitted

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go [hunk 3]
```diff
@@ -45,9 +46,11 @@ func (s *StringSlice) String() string {
 }
 
 func (s *StringSlice) Set(val string) error {
-	if s.value == nil || !s.changed {
-		v := make([]string, 0)
-		s.value = &v
+	if s.value == nil {
+		return fmt.Errorf("no target (nil pointer to []string)")
```
Comment (lavalamp) [inline staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]: IMO it's better to make the `NewStringSlice` detect this (and e.g. panic)? This is really a coding error, not a runtime error, you want to detect it as soon as possible. It'd be best if you could make it not compile, but I don't see a good way to do that.

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go [hunk 4]
```diff
@@ -45,9 +46,11 @@ func (s *StringSlice) String() string {
...
+	}
+	if *s.value == nil || !s.changed {
+		*s.value = make([]string, 0)
```
Comment (lavalamp) [inline staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:53]: append already does this if the first arg is nil, so I don't think you need the first clause in the `if` at least

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go (hunk 3 above)
Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:50]:
```golang
func NewStringSlice(s *[]string) *StringSlice {
	if s == nil {
		panic("no target (nil pointer to []string)")
	}
	return &StringSlice{value: s}
}
```
Is this? This is somewhat inconsistent with the previous code style

File: staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go (hunk 4 above)
Comment (vazmin) [reply staging/src/k8s.io/component-base/cli/flag/string_slice_flag.go:53]: As far as I can think of, this is indeed redundant. But I paid attention to similar codes, they all wrote this way, so I kept the first clause in the `if`.

https://github.com/kubernetes/kubernetes/blob/48b7492f92997ac8cbb8b6c559a0c68cba3021ab/staging/src/k8s.io/component-base/cli/flag/map_string_string.go#L69

**Pull request merged**